File: Main/150/200.cc
#include <vector>
using namespace std;

// easy question -> dfs on every square that is one
class Solution {
  public:
    void f(vector<vector<char>> &grid, int r, int c) {
        if (r < 0 || c < 0 || r >= grid.size() || c >= grid[0].size()) return;
        if (grid[r][c] == '0') return;

        grid[r][c] = '0';
        // go all four directions
        f(grid, r + 1, c);
        f(grid, r - 1, c);
        f(grid, r, c + 1);
        f(grid, r, c - 1);
    }

    int numIslands(vector<vector<char>> &grid) {
        int count = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == '1') {
                    count++;
                    f(grid, i, j);
                }
            }
        }
        return count;
    }
};

File: src/archive/Zon/AmazonPrepOld/Graphs/Go/200.go
package graphs

func numIslands(grid [][]byte) int {
	count := 0
	for r := 0; r < len(grid); r++ {
		for c := 0; c < len(grid[0]); c++ {
			if grid[r][c] == '1' {
				count++
				dfs(grid, r, c)
			}
		}
	}
	return count
}

func dfs(grid [][]byte, r int, c int) {
	if r < 0 || r >= len(grid) || c < 0 || c >= len(grid[0]) || grid[r][c] == '0' {
		return
	}

	grid[r][c] = '0'
	dfs(grid, r+1, c)
	dfs(grid, r-1, c)
	dfs(grid, r, c+1)
	dfs(grid, r, c-1)
}
